Implement an extension point deprecation strategy

LeonarddeR opened the pull request:
### Link to issue number:
Follow up of https://github.com/nvaccess/nvda/pull/17011
Slightly related to #17635, as this oversight was found there

### Summary of the issue:
In https://github.com/nvaccess/nvda/pull/17011, `braille.filter_displaySize` was deprecated, but there was not yet a mechanism in place that logs warnings about usage of the deprecated extension point.

### Description of user facing changes
Warnings in the log when `braille.filter_displaySize.register` is called.

### Description of development approach
1. Added an optional keyword argument _deprecationmessage to HandlerRegistrar. If set, it will be logged when a handler is registered.
2. Removed remaining handlers in core, changed them into handlers for `filter_displayDimensions`
3. Unit test expansion, missing tests for `filter_displayDimensions`

### Testing strategy:
Tested that a warning is logged when registering 

### Known issues with pull request:
None known

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

<!-- Please keep the following -->
@coderabbitai summary

LeonarddeR committed:
Implement deprecation for extension points

LeonarddeR committed:
Tests

LeonarddeR committed:
Rename

SaschaCowley committed:
Fix SAPI 4 driver (#17599)

Fixes #17516

Summary of the issue:
After the move to exclusively Windows core audio APIs, the SAPI4 driver stopped working.

Description of user facing changes
The SAPI4 driver works again.

A warning is shown the first time the user uses SAPI4 informing them that it is deprecated.

Description of development approach
Implemented a function to translate between MMDevice Endpoint IDs and WaveOut device IDs, based on <https://learn.microsoft.com/en-gb/windows/win32/coreaudio/device-roles-for-legacy-windows-multimedia-applications>.

Added a config key, `speech.hasSapi4WarningBeenShown`, which defaults to False.
Added a synthChanged callback that shows a dialog when the synth is set to SAPI4 if this config key is False and this is not a fallback synthesizer.

Testing strategy:
Ran NVDA, from source and installed, and on the user desktop, in secure mode, and on secure desktops, and used it with SAPI4. Changed the audio output device to ensure audio was routed as expected.

Known issues with pull request:
When first updating to a version with this PR merged, if the user uses SAPI4 as their primary speech synth, they will be warned about its deprecation in the launcher and when they first start the newly updated NVDA. This is unavoidable as we don't save config from the launcher.

The dialog is only shown once per config profile, so may be missed by some users.
Other options I have considered include:

* Making this a nag dialog that appears, say, once a week or once a month.
* Also making a dialog appear whenever the user manually sets their synth to SAPI4.
* Adding a new dialog in 2025.4 (or the last release before 2026.1) that warns users that this will be the last release to support SAPI4.
* Adding a dialog when updating to 2026.1 that warns users that they will no longer be able to use SAPI4.
* Adding a Windows toast notification that appears every time NVDA starts with SAPI4 as the synth.

The warning dialog is shown after SAPI4 is loaded. In the instance that the user is already using SAPI4, this is correct behaviour. In the case of switching to SAPI4, perhaps a dialog should appear before we terminate the current synth and initialise SAPI4.

SaschaCowley committed:
Fix up config upgrade steps from schema version 14 to 15 (#17642)

Fixes #17637
Follow-up to #17505

Summary of the issue:
The config upgrade steps for moving from version 14 to 15 of the config schema may leave the user's config in an invalid state if either of `["keyboard"]["speakTypedCharacters"]` or `["keyboard"]["speakTypedWords"]` is not a boolean.
This is because the upgrade steps, as implemented, take no action if those values are not bools.
For most upgrade steps, this approach is perfectly valid.
However, version 15 of the schema expects these values to be integers, so validation of the upgraded config will fail.

Description of user facing changes
The configuration upgrade should not corrupt the user's config.

Description of development approach
Instead of just `return`ing when we get a `ValueError` in `upgradeConfigFrom_14_to_15`, delete the offending config entry.


Testing strategy:
Tested by creating valid and invalid config strings, instantiating `ConfigObj`s with them, and feeding those to `upgradeConfigFrom_14_to_15`.

Known issues with pull request:
A user may lose their setting for "Speak typed characters" or "Speak typed words", if configobj doesn't recognise it as a boolean.
Theoretically this shouldn't be possible, as the config wouldn't have validated under any of the prior schema versions.
Nevertheless this seems to have happened, and it is better to lose (at most) 2 settings than all settings.

LeonarddeR committed:
Fix EcoBraille and Lilly braille display drivers (#17635)

Fixes #17502

Summary of the issue:
In #15386, support was added for numRows and numCols on braille displays. Setting numCells explicitly would also set numCols.
In #17011, we're much more relying on numCols and numRows.

For the eco braille and Lilli drivers, numCells was implemented by overriding _get_numCells. However this way, the setter of numCells would never be called and therefore numCols would always be 0.

Description of user facing changes
Fix braille for Eco Braille and Lilli displays.

Description of development approach
When calculating display dimensions on the braille handler, when display.numRows is 1, use display.numCells instead of display.numCols. This accounts for cases where numCells is implemented and numRows/numCols are not.
Worked around the ZeroDevisionError when getting windowStartPost on a buffer
Initialize _windowRowBufferOffsets to [(0, 0)] instead of [(0, 1)]. The latter suggests that the window contains one cell, which is not the case. @michaelDCurran you might want to check this.

lmacka committed:
Integrate new AppVeyor webhook (#17663)

Soft-integration of new Appveyor webhook which should run after the normal build on zebi in a non-blocking fashion.

lmacka committed:
Update webhook (#17668)

Try using the deploy webhook the AppVeyor way; we should be able to pass those variable in the environment instead of setting up a custom deploy pipeline.
Using the native way allows deploys to be retried via the UI, which is very helpful.
https://www.appveyor.com/docs/deployment/webhook/

SaschaCowley committed:
Add a payload argument to message dialog callbacks (#17660)

Fixes #17658 

Summary of the issue:
There is currently no way to pass information between message dialogs and their callbacks. This reduces the functionality available to callbacks in future.

Description of user facing changes
None.

Description of development approach
Add a `Payload` dataclass to `gui.message`. It currently has no fields, though these can be added without breaking the API in future, for example to implement #17646 .
In `MessageDialog._executeCommand`, if a command has a callback, instantiate a `Payload` and pass it to the callback when calling it.
Update the few uses of the message dialog API in NVDA which use callbacks:

* `synthDrivers.sapi4._sapi4DeprecationWarning`

Updated the developer guide to note the new requirements for callback functions.

Testing strategy:
Ran NVDA from source, and triggered the affected dialogs. Ensured they worked as expected.

Ran from source, with `versionInfo.updateVersionType` set to `"snapshot:alpha`, and checked that updating worked.

Built and installed this branch, then ran the CRFT, cancelling and continuing, and ensured that no issues occured.

Known issues with pull request:
None known.

SaschaCowley commented on tests/unit/__init__.py:
```diff
@@ -100,13 +100,13 @@
-	return 40
+# For braille unit tests, we need to enable the braille handler by providing it display dimensions
+# Give the display one row with 40 cells
+def getFakeDisplayDimensions(dimensions: braille.DisplayDimensions) -> int:
```
```suggestion
def getFakeDisplayDimensions(dimensions: braille.DisplayDimensions) -> braille.DisplayDimensions:
```

SaschaCowley commented on source/brailleViewer/__init__.py:
Please update copyright header

SaschaCowley commented on source/extensionPoints/util.py:
Please update copyright header

SaschaCowley commented on tests/system/libraries/SystemTestSpy/speechSpyGlobalPlugin.py:
Please update copyright

SaschaCowley commented on tests/unit/__init__.py:
Please update copyright

SaschaCowley reviewed (changes requested):
Looks pretty good, just a couple of minor changes

seanbudd committed:
Delete outdated translated copies of developer guide (#17641)

Closes #13333

Summary of the issue:
There are two unmaintained files, which are translated versions of the developer guide.
These have been unmaintained for over a decade.

Description of user facing changes
None

Description of development approach
Remove files

michaelDCurran committed:
appveyor.yml: don't limit deploy with rules. Syntax was wrong and it is not needed. (#17672)

CyrilleB79 commented on source/extensionPoints/util.py:
```diff
@@ -115,6 +123,11 @@ def register(self, handler: HandlerT):
 				raise TypeError("Registering unbound instance methods not supported.")
+		if self._deprecationMessage:
+			if NVDAState._allowDeprecatedAPI():
+				log.warning(self._deprecationMessage, stack_info=True)
```
@SaschaCowley, for deprecation we usually use `log.warning` in NVDA code base.

Here, you suggest to use `warnings.warn` instead. Is this suggestion applicable to this specific case only or to all deprecation codes in NVDA? If the latter, I'd suggest to discuss this and handle this in a specific issue / PR. If you suggest to use `warnings.warn` in the current specific case, could you explain why it differs from other deprecation code in NVDA?

Without any other explanation, I prefer to stick to the current use of `log.warning` using `stack_info=True` kwarg, since this give an indication to add-on authors where the deprecated code needs to be modified.

SaschaCowley replied:
My preference would be to switch to using `warnings.warn(msg, DeprecationWarning`) everywhere. That would allow us the ability to further customise the way deprecation warnings are shown in future - for instance, we could make unit tests fail on deprecation warnings. Note that we can fairly easily add traceback information to warnings, though we don't currently do so when we log them. We could also potentially add a helper method to automatically call `warnings.warn` with `category=DebugWarning` and `stacklevel=1` (to avoid warning in the deprecated method itself), and to exclude non core files when we update to python>=3.12 with the `skip_file_prefixes` argument. @seanbudd what are your thoughts? Is this worth opening an issue/pr about? Is it worth doing here?

LeonarddeR replied:
@seanbudd Could you give your preference here?

seanbudd replied:
I think we should use both or only logging. It really depends on the context.
`warnings.warn` only ever shows once, and we want it to show every time the API is called.
Unless there is a case where we only want the warning to show once, we should use logging.

SaschaCowley replied:
The behaviour of only emitting a warning once can be customised. See [The Warnings Filter in the documentation for warnings](https://docs.python.org/3/library/warnings.html#the-warnings-filter)

seanbudd replied:
Ah if that's the case I think we should move over to using it.
However, I think we need a separate PR first to make improve the behaviour of `warnings.warn`, and to update [deprecations.md](https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/deprecations.md).
I'd consider this out of scope of this PR.

LeonarddeR committed:
Apply suggestions from code review

Co-authored-by: Sascha Cowley <16543535+SaschaCowley@users.noreply.github.com>

pre-commit-ci[bot] committed:
Pre-commit auto-fix

LeonarddeR committed:
Review actions

LeonarddeR commented on source/braille.py:
```diff
@@ -2289,7 +2289,9 @@ def formatCellsForLog(cells: List[int]) -> str:
+filter_displaySize = extensionPoints.Filter[int](
+	_deprecationMessage="braille.filter_displaySize is deprecated. Use braille.filter_displayDimensions instead.",
+)
 """
```
There was a deprecaation notice at the bottom of the doc string, I moved it to the top.

SaschaCowley commented:
@LeonarddeR looks like something's gone wrong when merging in master

LeonarddeR committed:
Merge remote-tracking branch 'origin/master' into deprecateEp

seanbudd commented on source/brailleViewer/__init__.py:
```diff
@@ -1,14 +1,18 @@
 # This file is covered by the GNU General Public License.
 # See the file COPYING for more details.
-from typing import Optional
+import typing
```
```suggestion

import typing
```

seanbudd commented on tests/unit/extensionPointTestHelpers.py:
```diff
@@ -1,7 +1,7 @@
 # This file is covered by the GNU General Public License.
 # See the file COPYING for more details.
-# Copyright (C) 2023 NV Access Limited, Babbage B.V., Leonard de Ruijter
+# Copyright (C) 2023, 2025 NV Access Limited, Babbage B.V., Leonard de Ruijter
```
```suggestion
# Copyright (C) 2023-2025 NV Access Limited, Babbage B.V., Leonard de Ruijter
```

seanbudd committed:
Update source/brailleViewer/__init__.py

seanbudd committed:
Update tests/unit/extensionPointTestHelpers.py
